Improved mount detection using openat2 for kernel 5.10+

manugupt1 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind cleanup

#### What this PR does / why we need it:

Faster mount detection for linux kernel 5.10+ using openat2 speeding up pod churn rates.


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes # https://github.com/kubernetes/kubernetes/issues/105642

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Faster mount detection for linux kernel 5.10+ using openat2 speeding up pod churn rates. On Kernel versions less 5.10, it will fallback to using the original way of detecting mount points i.e by parsing /proc/mounts.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

manugupt1 commented:
/kind cleanup

jsafrane commented:
/ok-to-test

jingxu97 commented:
Can this mountedFast function is also used by IsNotMountPoint? Is it work for bind mount too?

manugupt1 commented:
> Can this mountedFast function is also used by IsNotMountPoint? Is it work for bind mount too?

Only for kernel 5.10+ / when openat2 is supported: https://github.com/moby/sys/blob/main/mountinfo/mounted_linux_test.go#L303-L320

jsafrane commented on staging/src/k8s.io/mount-utils/mount_unsupported.go:
```diff
@@ -86,3 +88,49 @@ func (mounter *SafeFormatAndMount) formatAndMountSensitive(source string, target
+// IsNotMountPoint enumerates all the mountpoints using List() and
+// the list of mountpoints may be large, then it uses
+// isMountPointMatch to evaluate whether the directory is a mountpoint.
+func IsNotMountPoint(mounter Interface, file string) (bool, error) {
```
This is `mount_unsupported.go`, I think returning `errUnsupported` is perfectly fine here.

jsafrane commented on staging/src/k8s.io/mount-utils/mount.go:
```diff
@@ -220,52 +219,6 @@ func GetDeviceNameFromMount(mounter Interface, mountPath string) (string, int, e
-		return notMnt, nil
-	}
-
-	// Resolve any symlinks in file, kernel would do the same and use the resolved path in /proc/mounts.
-	resolvedFile, err := filepath.EvalSymlinks(file)
-	if err != nil {
-		return true, err
-	}
-
-	// check all mountpoints since IsLikelyNotMountPoint
-	// is not reliable for some mountpoint types.
-	mountPoints, mountPointsErr := mounter.List()
-	if mountPointsErr != nil {
-		return notMnt, mountPointsErr
-	}
-	for _, mp := range mountPoints {
-		if isMountPointMatch(mp, resolvedFile) {
-			notMnt = false
-			break
-		}
-	}
```
Looking at the copies of the code in this PR, I think it would make sense to create a common function (I am bad at naming, `isNotMountPointProc` sounds bad) and call it when `MountedFast`/`IsLikelyNotMountPoint` is not sure on Linux/Windows

manugupt1 commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -637,6 +640,68 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+func IsNotMountPoint(mounter Interface, file string) (bool, error) {
+	klog.Errorf("IsNotMountPoint: file: %s detection started", file)
+	isMnt, sure, err := mountinfo.MountedFast(file)
+	klog.Errorf("IsNotMountPoint: file: %s mounted fast returned: %t, %t, %w", file, isMnt, sure, err)
+	if sure == true && err == nil {
+		klog.Errorf("IsNotMountPoint: file: %s returning !isMnt: %t, nil (returning)", file, !isMnt)
+		return !isMnt, nil
+	}
+	var perr = &fs.PathError{}
+	if err != nil && errors.As(err, perr) {
+		err = perr.Err
+		klog.Errorf("IsNotMountPoint: file: %s, found path error: %w; extracting original: %w", file, perr, err)
+	} else {
+		klog.Errorf("IsNotMountPoint: file: %s, not a path error but error is : %w", file, err)
+	}
```
Opened PR: https://github.com/moby/sys/pull/109

manugupt1 commented:
/test pull-kubernetes-node-e2e-containerd
/test pull-kubernetes-e2e-aks-engine-windows-containerd
/test pull-kubernetes-e2e-kind-ipv6 —

manugupt1 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

manugupt1 commented:
/test pull-kubernetes-node-e2e-containerd

manugupt1 commented:
/kind bug

kolyshkin commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -637,6 +638,47 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+		return notMnt, isMntErr
+	}
+	// Resolve any symlinks in file, kernel would do the same and use the resolved path in /proc/mounts.
+	resolvedFile, err := filepath.EvalSymlinks(file)
```
nit: what if a path is not absolute?

manugupt1 replied:
It will resolve against from the current directory. Eg: https://go.dev/play/p/9CqUNqev4yX

kolyshkin commented on staging/src/k8s.io/mount-utils/mount_unsupported.go:
```diff
@@ -86,3 +88,49 @@ func (mounter *SafeFormatAndMount) formatAndMountSensitive(source string, target
+		return notMnt, notMntErr
+	}
+	// identified as mountpoint, so return this fact.
+	if notMnt == false {
```
nit: `if !notMnt`.

Which also brings us to a different issue. Can we rename `IsNotMountPoint` to `IsMountPoint`?

manugupt1 replied:
I am actually not sure to rename as we already have `IsLikelyNotMountPoint`. I am okay either way.

kolyshkin commented:
@manugupt1 while reviewing this, I can't help but notice the first 3 commits should be squashed (for the sake of cleaner git history, and also for the sake of reviewers).

manugupt1 commented:
/test pull-kubernetes-e2e-kind-ipv6

manugupt1 committed:
Use faster mountdetections using openat2 for IsNotMountPoint

Reimplement IsNotMountpoint as IsMountPoint and return appropriate
error when a mount point is not detected.

IsMountPoint depends on MountedFast function in moby/sys/mountinfo
that uses openat2 call in kernel versions5.10 to detect
mount points without falling back to /proc/mounts.

Mark IsNotMountPoint as deprecated in order to ask the users to
use IsMountPoint directly.

manugupt1 committed:
Fix type in mount_windows.go

manugupt1 commented:
/test pull-kubernetes-integration

manugupt1 commented:
/test pull-kubernetes-integration

kolyshkin commented on go.mod:
```diff
@@ -56,7 +56,6 @@ require (
 	github.com/lithammer/dedent v1.1.0
 	github.com/lpabon/godbc v0.1.1 // indirect
 	github.com/moby/ipvs v1.0.1
-	github.com/moby/sys/mountinfo v0.6.0 // indirect
```
Hmm, this hunk does not look good. It should be something like
```diff
-       github.com/moby/sys/mountinfo v0.6.0 // indirect
+       github.com/moby/sys/mountinfo v0.6.1 // indirect
```

I have checked vendoring tools and they work fine.

What you need to create this commit is
```console
$ ./hack/pin-dependency.sh github.com/moby/sys/mountinfo v0.6.1
$ ./hack/update-vendor.sh
$ ./hack/lint-dependencies.sh 
```

(and the usual `git` commands)

manugupt1 replied:
I did exactly what you told.

```
  kubernetes git:(improve-mount-detection) ./hack/pin-dependency.sh github.com/moby/sys/mountinfo v0.6.1
Running: go mod download github.com/moby/sys/mountinfo@v0.6.1
Resolved to github.com/moby/sys/mountinfo@v0.6.1
Running: go mod edit -require github.com/moby/sys/mountinfo@v0.6.1
Running: go mod edit -replace github.com/moby/sys/mountinfo=github.com/moby/sys/mountinfo@v0.6.1

Run hack/update-vendor.sh to rebuild the vendor directory
➜  kubernetes git:(improve-mount-detection) ✗ ./hack/update-vendor.sh                                      
+++ [0425 17:40:58] logfile at /tmp/update-vendor.QxrQ/update-vendor.log
+++ [0425 17:40:58] go.mod: update staging references
+++ [0425 17:41:03] go.mod: propagate to staging modules
+++ [0425 17:41:03] go.mod: sorting staging modules
+++ [0425 17:41:18] go.mod: tidying
+++ [0425 17:41:32] go.mod: adding generated comments
+++ [0425 17:41:33] vendor: updating internal modules
+++ [0425 17:41:33] vendor: running 'go mod vendor'
+++ [0425 17:41:34] vendor: updating vendor/LICENSES
+++ [0425 17:41:39] vendor: creating OWNERS file
+++ [0425 17:41:39] NOTE: don't forget to handle vendor/* files that were added or removed
➜  kubernetes git:(improve-mount-detection) ./hack/lint-dependencies.sh          
Check all unwanted dependencies and update its status.
All pinned versions of checked dependencies match their preferred version.
➜  kubernetes git:(improve-mount-detection) git status
On branch improve-mount-detection
Your branch is up to date with 'origin/improve-mount-detection'.

nothing to commit, working tree clean
➜  kubernetes git:(improve-mount-detection) git log --oneline -n 3                                       
32fe234004b (HEAD -> improve-mount-detection, origin/improve-mount-detection) Fix type in mount_windows.go
0097b228bff Use faster mountdetections using openat2 for IsNotMountPoint
c625fb38ac1 Update mountinfo to v0.6.1.
➜  kubernetes git:(improve-mount-detection) 
  kubernetes git:(improve-mount-detection) ./hack/pin-dependency.sh github.com/moby/sys/mountinfo v0.6.1
Running: go mod download github.com/moby/sys/mountinfo@v0.6.1
Resolved to github.com/moby/sys/mountinfo@v0.6.1
Running: go mod edit -require github.com/moby/sys/mountinfo@v0.6.1
Running: go mod edit -replace github.com/moby/sys/mountinfo=github.com/moby/sys/mountinfo@v0.6.1

Run hack/update-vendor.sh to rebuild the vendor directory
➜  kubernetes git:(improve-mount-detection) ✗ ./hack/update-vendor.sh                                      
+++ [0425 17:40:58] logfile at /tmp/update-vendor.QxrQ/update-vendor.log
+++ [0425 17:40:58] go.mod: update staging references
+++ [0425 17:41:03] go.mod: propagate to staging modules
+++ [0425 17:41:03] go.mod: sorting staging modules
+++ [0425 17:41:18] go.mod: tidying
+++ [0425 17:41:32] go.mod: adding generated comments
+++ [0425 17:41:33] vendor: updating internal modules
+++ [0425 17:41:33] vendor: running 'go mod vendor'
+++ [0425 17:41:34] vendor: updating vendor/LICENSES
+++ [0425 17:41:39] vendor: creating OWNERS file
+++ [0425 17:41:39] NOTE: don't forget to handle vendor/* files that were added or removed
➜  kubernetes git:(improve-mount-detection) ./hack/lint-dependencies.sh          
Check all unwanted dependencies and update its status.
All pinned versions of checked dependencies match their preferred version.
➜  kubernetes git:(improve-mount-detection) git status
On branch improve-mount-detection
Your branch is up to date with 'origin/improve-mount-detection'.

nothing to commit, working tree clean
➜  kubernetes git:(improve-mount-detection) git log --oneline -n 3                                       
32fe234004b (HEAD -> improve-mount-detection, origin/improve-mount-detection) Fix type in mount_windows.go
0097b228bff Use faster mountdetections using openat2 for IsNotMountPoint
c625fb38ac1 Update mountinfo to v0.6.1.
➜  kubernetes git:(improve-mount-detection) 
  kubernetes git:(improve-mount-detection) ./hack/pin-dependency.sh github.com/moby/sys/mountinfo v0.6.1
Running: go mod download github.com/moby/sys/mountinfo@v0.6.1
Resolved to github.com/moby/sys/mountinfo@v0.6.1
Running: go mod edit -require github.com/moby/sys/mountinfo@v0.6.1
Running: go mod edit -replace github.com/moby/sys/mountinfo=github.com/moby/sys/mountinfo@v0.6.1

Run hack/update-vendor.sh to rebuild the vendor directory
➜  kubernetes git:(improve-mount-detection) ✗ ./hack/update-vendor.sh                                      
+++ [0425 17:40:58] logfile at /tmp/update-vendor.QxrQ/update-vendor.log
+++ [0425 17:40:58] go.mod: update staging references
+++ [0425 17:41:03] go.mod: propagate to staging modules
+++ [0425 17:41:03] go.mod: sorting staging modules
+++ [0425 17:41:18] go.mod: tidying
+++ [0425 17:41:32] go.mod: adding generated comments
+++ [0425 17:41:33] vendor: updating internal modules
+++ [0425 17:41:33] vendor: running 'go mod vendor'
+++ [0425 17:41:34] vendor: updating vendor/LICENSES
+++ [0425 17:41:39] vendor: creating OWNERS file
+++ [0425 17:41:39] NOTE: don't forget to handle vendor/* files that were added or removed
➜  kubernetes git:(improve-mount-detection) ./hack/lint-dependencies.sh          
Check all unwanted dependencies and update its status.
All pinned versions of checked dependencies match their preferred version.
➜  kubernetes git:(improve-mount-detection) git status
On branch improve-mount-detection
Your branch is up to date with 'origin/improve-mount-detection'.

nothing to commit, working tree clean
➜  kubernetes git:(improve-mount-detection) git log --oneline -n 3                                       
32fe234004b (HEAD -> improve-mount-detection, origin/improve-mount-detection) Fix type in mount_windows.go
0097b228bff Use faster mountdetections using openat2 for IsNotMountPoint
c625fb38ac1 Update mountinfo to v0.6.1.
➜  kubernetes git:(improve-mount-detection) 
```

I can see no changes.

manugupt1 replied:
Hmm! not sure what happened; doing this from master gave the output that you gave. I will resolve this.

kolyshkin commented on vendor/modules.txt:
```diff
@@ -518,8 +518,7 @@ github.com/moby/ipvs
 github.com/moby/spdystream/spdy
-# github.com/moby/sys/mountinfo v0.6.0 => github.com/moby/sys/mountinfo v0.6.0
-## explicit
+# github.com/moby/sys/mountinfo v0.6.1 => github.com/moby/sys/mountinfo v0.6.1
```
This part also looks wrong, the `explicit` line should not be removed.

manugupt1 replied:
[e65ce59d421ca705c853619153a8cbc4a880c1d7](https://github.com/kubernetes/kubernetes/pull/109217/commits/57cb3264320f3616e319e4b6af5e142a46913632)

https://github.com/kubernetes/kubernetes/pull/109217/commits/13a66757f6e9ae6d45fd34762ad973be858297b3

These two commits should be squashed as with introduction of this change, its not an indirect dependency anymore.

manugupt1 commented:
/test pull-kubernetes-e2e-aks-engine-windows-containerd

jsafrane commented on staging/src/k8s.io/mount-utils/mount.go:
```diff
@@ -187,6 +194,27 @@ func getMountRefsByDev(mounter Interface, mountPath string) ([]string, error) {
+// It should return ErrNotExist when the directory does not exist.
+// IsNotMountPoint is more expensive than IsLikelyNotMountPoint.
+// IsNotMountPoint detects bind mounts in linux.
+// IsNotMountPoint enumerates all the mountpoints using List() and
+// the list of mountpoints may be large, then it uses
+// isMountPointMatch to evaluate whether the directory is a mountpoint.
```
It would be better to mention that `IsNotMountPoint *may* enumerate all the mountpoints using List()...`

jsafrane commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -637,6 +638,46 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+		return isMnt, nil
+	}
+	if isMntErr != nil && os.IsPermission(isMntErr) {
+		// We were not allowed to do the simple stat() check, e.g. on NFS with
+		// root_squash. Fall back to /proc/mounts check below.
+		return false, nil
```
The comment and the code does not match. There is no fall back to /proc/mounts here.

jingxu97 commented on staging/src/k8s.io/mount-utils/mount.go:
```diff
@@ -66,6 +65,14 @@ type Interface interface {
+	// IsMountPoint may enumerate all the mountpoints using List() and
+	// the list of mountpoints may be large, then it uses
+	// isMountPointMatch to evaluate whether the directory is a mountpoint.
+	IsMountPoint(file string) (bool, error)
```
why adding IsMountPoint func? Is it used by any code?

manugupt1 replied:
I was addressing this comment: https://github.com/kubernetes/kubernetes/pull/109217#discussion_r850410452

jingxu97 commented on staging/src/k8s.io/mount-utils/mount.go:
```diff
@@ -187,6 +194,27 @@ func getMountRefsByDev(mounter Interface, mountPath string) ([]string, error) {
+// IsNotMountPoint detects bind mounts in linux.
+// IsNotMountPoint may enumerate all the mountpoints using List() and
+// the list of mountpoints may be large, then it uses
+// isMountPointMatch to evaluate whether the directory is a mountpoint.
```
maybe update the comment a little bit?

jingxu97 commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -637,6 +638,46 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+// IsMountPoint determines if a directory is a mountpoint.
+// It should return ErrNotExist when the directory does not exist.
+// IsMountPoint is more expensive than IsLikelyNotMountPoint.
+// IsMountPoint detects bind mounts in linux.
```
update the comment to match the code

manugupt1 replied:
How does this sound?

```
IsMountPoint first tries to detects bind mounts and any other mounts in linux by using MountedFast function. If the MountedFast function returns sure as true and err as nil, a mount is detected successfully. In other cases, bind mount is detected by enumerating all the mountpoints using List()....
```

jingxu97 commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -637,6 +638,46 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+// the list of mountpoints may be large, then it uses
+// isMountPointMatch to evaluate whether the directory is a mountpoint.
+func (mounter *Mounter) IsMountPoint(file string) (bool, error) {
+	isMnt, sure, isMntErr := mountinfo.MountedFast(file)
```
before calling MountedFast, should we still call IsLikelyNotMountPoint like before so it can solve certain cases first.

Also if file does not exist, it will return immediately instead of executing the following.

manugupt1 replied:
MountedFast already uses stat function internally. I would like to keep this as it is backed by actual tests in the moby/sys CI; so bug fixes in the upstream library can be shipped with more confidence.

This change is backward compatible when file does not exist, it returns an error ErrNotExist and the user is supposed to determine what to do with it. This is where: https://github.com/kubernetes/kubernetes/pull/109217/files#diff-5a53703b1baddb2caa650077f26075c59361575c84387e1e9ad5570a738dd473L338-L341, ErrNotExist is returned. This is the comment (https://github.com/kubernetes/kubernetes/pull/109217/files#diff-e0ed3eecf397212be5c090559125652c439c5b93699be999c4bf7e06ae6531a2L224) that mentions ErrNotExist should be handled by the caller.

jingxu97 replied:
the the first link the correct one you mentioned? Just make sure mountinfo.MountedFast(file) returns correct error if file does not exit. Looks like it is using open sys call.

manugupt1 replied:
Technically, MountedFast calls stat first, but it works in all cases: https://go.dev/play/p/GyieSQgECO-.

jingxu97 replied:
Before IsNotMountPoint() function implementation, it calls isLikelyNotMountPoint first. Is there any downside of using/not using this function first?

manugupt1 replied:
Really good question; mountedFast has the same logic: 
MountedFast first normalizes the path and evaluates all the symlinks first. Lstat is used (instead of stat in case of IsLikelyNotMountPoint). With Lstat, symlinks are not followed; however since the path is already normalized it is not required.

The other part of the logic is the same: https://github.com/moby/sys/blob/523ebb9c6cc361d558aaf903946709d78efa0f6a/mountinfo/mounted_unix.go#L13-L31

manugupt1 commented:
/test pull-kubernetes-unit

The unit test seems like a flake; I have ran the unit test over 1000 times and seems like there were previous failures.
Similar to: https://github.com/kubernetes/kubernetes/issues/108906. https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/109217/pull-kubernetes-unit/1529849597139095552

manugupt1 commented:
/test pull-kubernetes-unit

jingxu97 commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -637,6 +638,46 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+		return isMnt, nil
+	}
+	if isMntErr != nil && os.IsPermission(isMntErr) {
+		// We were not allowed to do the simple stat() check, e.g. on NFS with
```
if it is permission error, the following check should still work?

manugupt1 replied:
Yes. it will work: see: https://pkg.go.dev/os#IsPermission. This should not be the case though because of the following commit in moby: https://github.com/moby/sys/commit/925ece265955e861c1292597fe0da9dd1ed66668. Note that os.Stat returns Path Error, but os.IsPermission is capably of handling one level of error.

I improved this case now so that we also account for error wrapping as well for os.Permission so that all error is properly unwrapped.

jingxu97 replied:
how about the following to save a few lines?

```
if isMntErr !=nil && !errors.Is(isMntErr, fs.ErrPermission) {
    return false, err
}

```

jingxu97 replied:
also as @liggitt, we should unwrap error before returning it here?

manugupt1 replied:
I just addressed the two feedback pieces.

manugupt1 replied:
Found one more place and addressed it as well: https://github.com/kubernetes/kubernetes/pull/109217/commits/fe5f3ef8dffc47647b510e861df03cbb6a1c4290

Here is an example: https://github.com/kubernetes/kubernetes/pull/109217/commits/fe5f3ef8dffc47647b510e861df03cbb6a1c4290

jingxu97 replied:
How about using error.As() as suggested by @liggitt?

manugupt1 replied:
I believe @liggitt was suggesting to unwrap to PathError but we want to return ErrNotExist which is not typed. Ref: https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/internal/oserror/errors.go;l=15-16;drc=55590f3a2b89f001bcadf0df6eb2dde62618302b 

Therefore I used this way to ensure that in case of ErrNotExist, we just return the bare error.

manugupt1 replied:
As an example, please check this playground: https://go.dev/play/p/NpKmQnbhMLi where unix errors cannot be converted to PathErrors.

manugupt1 commented:
/test pull-kubernetes-e2e-gce-storage-slow

jingxu97 commented:
/lgtm
/approve

jingxu97 commented:
/lgtm

manugupt1 commented:
/assign @liggitt

For dependency updates

liggitt commented on vendor/github.com/moby/sys/mountinfo/mounted_unix.go:
```diff
@@ -33,13 +32,13 @@ func mountedByStat(path string) (bool, error) {
 
 func normalizePath(path string) (realPath string, err error) {
```
Why was this change required for this PR? if we're relying on the error returned from this library not being wrapped, that seems really fragile. Shouldn't we be using errors.As to get the ErrNotExist error if present?

liggitt replied:
(I don't object to picking up the dependency bump otherwise)

jingxu97 replied:
So the suggestion is to not bump the dependency here, instead fixing the tests where it checks notfound error to use error.As, right?

manugupt1 replied:
This change is required because k8s depends on os.IsExist and os.IsNotExist: https://github.com/kubernetes/kubernetes/blob/1c526a536f83ca7a0dd77f455f85c336bca7e801/pkg/volume/util/subpath/subpath_linux.go#L103-L107 instead of errors.Is(err, fs.ErrNotExist) && !errors.Is(err, fs.ErrNotExist)

We can modify the PR to include this change as well or do as a follow up.

jingxu97 replied:
> The function is documented as returning ErrNotExist. 

Are you referring function IsMountPoint() here which changed to add [MountedFast](https://github.com/moby/sys/blob/main/mountinfo/mounted_linux.go#L30) from the library, right? Yes, I think we can unwrap error there

liggitt replied:
> Are you referring function IsMountPoint() here which changed to add [MountedFast](https://github.com/moby/sys/blob/main/mountinfo/mounted_linux.go#L30) from the library, right? Yes, I think we can unwrap error there

if we're unwrapping there (and testing to verify unwrapping is working), do we still need this library bump? if not, can we drop the dependency change from this PR to make it clear that this PR doesn't depend on the `github.com/moby/sys/mountinfo` library returning unwrapped errors?

cartermckinnon commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -637,6 +640,57 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+
+	// check all mountpoints since MountedFast is not sure.
+	// is not reliable for some mountpoint types.
+	mountPoints, mountPointsErr := mounter.List()
+	if mountPointsErr != nil {
+		return false, mountPointsErr
+	}
+	for _, mp := range mountPoints {
+		if isMountPointMatch(mp, resolvedFile) {
+			return true, nil
+		}
+	}
+	return false, nil
```
Can we use `GetMounts` with a `SingleFileFilter` here, like `moby/sys/mountinfo` does?

The performance issue I'm trying to address in https://github.com/kubernetes/kubernetes/pull/109676 is caused by the repetitive reads of `/proc/mounts` at:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/mount-utils/mount_linux.go#L544

And ultimately, optimizing the mount point checks is a better solution than trying to skip them (like #109676 tries to).

manugupt1 replied:
We can't check that because unlike umount, we need to verify consistency in mountpoint check. GetMounts does not do a consistency check: https://github.com/moby/sys/blob/74ec3fe1325a016151d353bb6654d960042f9e96/mountinfo/mountinfo_linux.go#L19

because of that you will run into race condition with procfs; where a mountpoint may be incorrectly recognized. This is the main reason; we are trying to solve the problem in different ways.

cartermckinnon replied:
Apologies if this isn't the right place to ask, but I've crawled the git history and I'm still not understanding why the consistency check is necessary, particularly in the `CleanupMountPoint` code path.

Can you elaborate on the race condition(s)? Is there a race between `Mount` returning successfully and a respective entry in `/proc/mounts` existing? Or in the other direction; does a successful `Unmount` not result in the entry going away?

The consistency check seems like an intractable approach, because we're essentially requiring the system to become idle before we can perform further mounting operations.

manugupt1 replied:
There was a bug in the kernel where /proc/mounts is read page by page. when 2 pages are read, then the 1st page can have an older snapshot and the 2nd page can be of an older snapshot reading to an inconsistent view of the mount-points. The assumption is that if 2 entire reads of /proc/mounts is the same and the mount entry is not present, then mountpoint is not present. This was solved in kernel 5.8+, commit https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=9f6c61f96f2d97cbb5f7fa85607bc398f843ff0f explains it much better.

cartermckinnon replied:
Thanks for the background!

It seems like we could break out of the consistency check if the mount was found in `/proc/mounts`, because this bug would only result in a mount being incorrectly omitted from `/proc/mounts`. I think such a change could be beneficial in the [first mount existence check in `CleanupMountPoint`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/mount-utils/mount_helper_common.go#L90), because the happy path is when the `mountPath` *is* a mount point and is *not* a corrupted mount. I'll do some testing before drafting a PR -- and please let me know if this idea doesn't make sense 😅

manugupt1 replied:
In the past, sig-storage has not accepted changes like removing a consistent read; although finding a case where consistent read is not required will be an awesome optimization.

This is the case that I was thinking can lead to issues.

Let's say we are reading /proc/mounts and it has 6 mount points initially.
A single page read can return 3 mountpoints. Between 2 page reads, the cursor might have changed 
and return a false positive that a mount-point does not exist when it actually does; leading
to data integrity issue.

As an example:


Suppose a page can have 3 mountpoints and we have 2 pages.

First content

SNP1, Page 1 :MP1, MP2, MP3
SNP1, Page 2  :MP4, MP5, MP6


2nd content (after MP1 has been unmounted which can happen between 2 page reads)

SNP2, Page 1 MP2, MP3, MP4 (1st page read can read MP4)
SNP2, Page 2 MP5, MP6 


the file content may have

MP1, MP2, MP3, MP5, MP6. In this case, MP4 is a mountpoint but has not been read.

Consistent read only partially solves the problem by comparing 2 consecutive reads but not wholly though 
as something like this might still happen.

cartermckinnon replied:
So, when determining if something is a mountpoint or not, if we encounter it at any point while parsing `/proc/mounts`, then we can return early -- we know that it must be a mount point. The bug can't result in mountpoints being mistakenly *included*, only mistakenly *excluded*.

For `CleanupMountPoint`, this only benefits the check before `Unmount`, but should be a meaningful optimization for the average case. After `Unmount`, however, we still need a consistent read to be confident that the path is *not* a mountpoint.

Looking at the code, this would be a significant refactor, because the consistent read operation is isolated from any notion of the target path we're checking. It's also not necessary if we have #109676 , because the consistent read is avoided both before and after `Unmount`. (Though some users could still benefit from it if their system does not have the desired "not mounted" behavior).

I'll draft a PR when I have a few minutes. I think it's worth doing, because the consistent read will be necessary for the foreseeable future, as the bug is only fixed in very recent kernels.

manugupt1 replied:
I believe returning early works if you can make sure that you are not reading more than a single page.  AFAIK, strace can be really useful way to ensure that. If that is possible, code refactor is possible.

manugupt1 committed:
Return unwrapped ErrNotExist when EvalSymlink returns PathError

By default filepath.EvalSymlink returns PathError. When a file is
not found, we should unwrap it and return ErrNotExist as this
is what this function expects.

Similar to the comment at:
https://github.com/kubernetes/kubernetes/pull/109217#discussion_r896272206

manugupt1 commented:
/test pull-kubernetes-e2e-kind-ipv6

manugupt1 commented:
/test pull-kubernetes-e2e-kind-ipv6

manugupt1 commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

manugupt1 commented:
/milestone v1.25

jingxu97 commented:
/lgtm

manugupt1 commented:
/test pull-kubernetes-e2e-kind

manugupt1 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

manugupt1 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

manugupt1 commented:
@liggitt @jingxu97 I fixed the error unwrapping part:
* https://github.com/kubernetes/kubernetes/pull/109217/commits/066b9caa94d5374cd4caa298dcaf3b6dfb1d14be#diff-5a53703b1baddb2caa650077f26075c59361575c84387e1e9ad5570a738dd473R725-R728 and here 
* https://github.com/kubernetes/kubernetes/pull/109217/commits/066b9caa94d5374cd4caa298dcaf3b6dfb1d14be#diff-5a53703b1baddb2caa650077f26075c59361575c84387e1e9ad5570a738dd473R725-R727

encodeflush commented:
Hi @manugupt1 
Bug triage team here!
It looks like this PR has the following linked PRs:

* https://github.com/kubernetes/kubernetes/pull/109117

Just checking in to see if this is still on track for k8s 1.25.

manugupt1 commented:
Sure! Please check now. I have reverted to 0.6.0 which we need. All the conformance tests, kind tests,  unit tests are passing (which would have failed otherwise.). For some reason I am not able to comment on the code base.

cc @jingxu97 @liggitt

manugupt1 commented:
> * Avoid mount point check when cleaning mount points #109117

The PR that you mentioned has been merged via a different PR https://github.com/kubernetes/kubernetes/pull/109676

jingxu97 commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -692,6 +695,63 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+	// root_squash. Fall back to /proc/mounts check below when
+	// fs.ErrPermission is returned.
+	if isMntErr != nil {
+		if errors.Is(isMntErr, fs.ErrNotExist) {
```
also just to make sure mountinfo.MountedFast will return ErrNotExist if file/path does not exist?

looks like it returns &os.PathError{Op: "openat2", Path: dir, Err: err}, 
moby/sys@88ece65#diff-1233c1d7752b2c506417d72bf871879082263410c1ee1cf3c0021f0e47a5d4faR47

manugupt1 replied:
Yep! PathError adds more context to it. That is why we are unwrapping to return bare error:

Please see this example for unwrapping cases where we will get true when errors.Is(isMntError, fs.ErrNotExist)
https://go.dev/play/p/DBNWKT8Qi6A

jingxu97 commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -692,6 +695,63 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+	}
+	// We were not allowed to do the simple stat() check, e.g. on NFS with
+	// root_squash. Fall back to /proc/mounts check below when
+	// fs.ErrPermission is returned.
```
move the comment below?

jingxu97 commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -692,6 +695,63 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+		// root_squash. Fall back to /proc/mounts check below when
+		// fs.ErrPermission is returned.
+		if !errors.Is(isMntErr, fs.ErrPermission) {
+			return false, isMntErr
```
how about return fs.ErrPermission here too? Does it match the previous behavior ?

manugupt1 replied:
Yes! we need to fallback when fs.ErrPermission occurs for nfs and squash.

jingxu97 replied:
oh, sorry, actually I mean unwrap isMntErr. Is that needed?

manugupt1 replied:
Oh that should not be needed because of our documentation, we only promise to return ErrNoExist.

manugupt1 commented:
/test pull-kubernetes-verify-govet-levee

jingxu97 commented:
/lgtm

in release note, maybe mention that for old kernel version, it will fall back to the original way to check mount point.

liggitt commented:
/approve
for go.mod file update

manugupt1 commented:
/test pull-kubernetes-e2e-capz-windows-containerd

BenTheElder commented:
/triage accepted
/priority important-longterm
